Firmware/messages.c

```diff
@@ -18,36 +18,44 @@
 
 #include "base.h"
 
-void bp_message_write_buffer(unsigned long strptr, size_t length) {
-    size_t index;
+void bp_message_write_buffer(unsigned long strptr) {
     unsigned char tblpag_prev = TBLPAG;
+    unsigned char index = 0;
+    char ch;
 
-    for (index = 0; index < length; index++) {
+    while (1) {
         TBLPAG = (strptr >> 16) & 0xFF;
-        switch (index % 3) {
+        switch (index) {
             case 0:
-                UART1TX(__builtin_tblrdl(strptr) & 0xFF);
+                ch = __builtin_tblrdl(strptr) & 0xFF;
+                index++;
                 break;
 
             case 1:
-                UART1TX((__builtin_tblrdl(strptr) >> 8) & 0xFF);
+                ch = (__builtin_tblrdl(strptr) >> 8) & 0xFF;
+                index++;
                 break;
 
             case 2:
-                UART1TX(__builtin_tblrdh(strptr) & 0xFF);
+                ch = __builtin_tblrdh(strptr) & 0xFF;
                 strptr += 2;
+                index = 0;
                 break;
+        }
 
-            default:
-                break;
+        if (ch != '\0') {
```

Comment:
How about moving this comparison in the `while` block instead?  The resulting code would be slightly clearer, and maybe shorter by a few bytes too.

Alternatively, something like this may do as well:

```c
do {
	TBLPAG = (strptr >> 16) & 0xFF;
	switch (index) {
	case 0:
		ch = __builtin_tblrdl(strptr) & 0xFF;
		index++;
		break;

	case 1:
		ch = __builtin_tblrdl(HI8(strptr)) & 0xFF;
		index++;
		break;

	case 2:
		ch = __builtin_tblrdh(strptr) & 0xFF;
		strptr += 2;
		index = 0;
		break;

	default:
		break;
	}

	if (ch != '\0') {
		UART1TX(ch);
	}	
} while (ch != '\0');
```

Reply:
At least with optimizations disabled, the while loop generates slightly shorter code. The difference is very small, so I'm OK with either form.

Reply:
Weird, I would expect the compiler to be smart enough to re-use the `ch != '\0'` value and/or flag results for the `while` check and save some code, even in the un-optimised version.

Reply:
Without optimizations, a variable will be re-read from memory every time it's referenced. -O0 isn't really meant to be used beyond diagnosing compiler issues, -O1 is really the sane minimum, and I was going to suggest moving to it as a default after the 7.1 release. In addition to the performance and code size improvements, it would also allow for eg. using constant folding and dead code elimination to reduce the amount of ifdef spaghetti.

Reply:
Fair point, I'll take your suggestion in and make -O1 the default later on.